### DISCUSSION TITLE
Target-typing new() when used next to a binary operator

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5561] [Language Ideas]
Currently, it is not possible to have an expression of the form `value + new(...)`, since `new()` does not target-type, despite there being enough context.

The proposal is to allow `new()` to target the expected type for the respective hand of the binary operator, if there is only one, disregarding implicit type conversions. It does sound hacky, working around the potential implicit conversions.

## Example
Consider the following struct:
```csharp
public record struct S(int Value)
{
    public static S operator +(int left, S right) => new(left + right.Value);
    public static S operator +(S left, int right) => new(left.Value + right);
}
```

The following is currently invalid, even if one could easily imply that `new()` should infer that the type is `int`:
```csharp
Console.WriteLine(new S(5) + new());
```

The current context is the following:
- `S` has no implicit type conversions
- If `S` appears on the LHS of a binary `+`, the RHS is always `int`

This means that `new()` can only safely target-type to `int`.

## Further Expansions

If there's multiple types that can bind to the respective side of the operator, choose the one that contains a constructor matching the given argument types. This approach feels hacky, and might easily become counter-intuitive in several cases where types contain simmilar constructors, if not identical. Additionally, `params` and default parameter values are going to make life even more painful. Personally, I would only be in favor of this kind of expansion for **different** constructor parameter counts, and only allowing `params` if the constructors with `params` have a greater total parameter count than the rest. This case sounds too specific anyway.

## Alternative Considerations

Implicit type conversions could not be ignored, and prohibit target-typing `new()` to a type that has implicit type conversions.

**Comment by jnm2** [comment]
The thing that is missing from this proposal is motivation. I love target-typed new, and my initial feeling is that this ability is one I would never use because it would be too confusing.
